test/validation/api/packet/packet.c

```diff
@@ -929,13 +929,16 @@ static void _verify_headroom_shift(odp_packet_t *pkt,
 	CU_ASSERT_PTR_NOT_NULL(data);
 	if (extended) {
 		CU_ASSERT(rc >= 0);
-		if (shift >= 0) {
-			CU_ASSERT(odp_packet_seg_len(*pkt) == shift - room);
-		} else {
-			CU_ASSERT(odp_packet_headroom(*pkt) >=
-				  (uint32_t)abs(shift) - seg_data_len);
+		if (rc == 0) {
+			if (shift >= 0) {
+				CU_ASSERT(odp_packet_seg_len(*pkt) ==
+					  shift - room);
```

Comment:
rc should never be zero, so the assert that checks that it is >=0 could just check that rc is strictly positive. The preceding code calls extend_head only when there is not enough headroom in which case either odp_packet_head() must change or data must move. And similarly for the truncation case. We could then get rid of all the code that was made conditional on rc being zero as it is obviously dead code.

Btw, there is a bug in the condition which decides whether pull can be done instead of trunc. Only seg_data_len - 1 bytes may be pulled, not all the data bytes in the segment.

> CU_ASSERT(odp_packet_seg_len(*pkt) == shift - room);

This assert here assumes that a new segment was added in the packet and contains the data bytes that did not fit in the old head room. But if segmentation changes, we do not come here. And in this function, segmentation always changes when extended is true, so this is dead code now.

If the preceding code was different and we could come here without segmentation changing, then this assert would not make any sense at all.

There are similar issues in the tailroom shift code.

Reply:
>rc should never be zero

Agree, (rc == 0) only if extend_head becomes equivalent to push_head which will never happen here. Will remove the dead code.

> Btw, there is a bug in the condition which decides whether pull can be done instead of trunc. Only seg_data_len - 1 bytes may be pulled, not all the data bytes in the segment.

This is addressed in #1266

Reply:
@JannePeltonen  On another thought, just for extend tail case, I think rc can be 0. Because we can achieve extend tail by conserving all existing pointers same except odp_packet_tail().

odp_packet_tail() cannot be conserved in anyway in any case when doing extend tail.

Reply:
> On another thought, just for extend tail case, I think rc can be 0. Because we can achieve extend tail beyond tailroom by conserving all existing pointers except odp_packet_tail().

I agree. The implementation could just make the segment larger if it had means to do so. But in that case the assert in the rc==0 case is incorrect as it assumes that a new segment was created.

Actually, I think that even trunc_tail might return 0 even if it has to remove the last segment. If the various packet pointers are not within the truncated range, they do not need to change. And if they are in the truncated range, then they should not be dereferenced anymore anyway, even in the pull_tail case, I suppose. I am not sure if the API is clear enough on the exact semantics.

Reply:
Thinking in that lines, even extend/trunc head can be done with by making the segment larger. Only odp_packet_head() pointer needs to change there.

Reply:
> Only odp_packet_head() pointer needs to change there.

But zero return code means that all pointers stay valid.

Reply:
You are right. 

I was trying to compare/correlate  trunc/extend head to trunc/extend tail semantics.  Like if trunc/extend tail can return 0 with change in odp_packet_tail(), then trunc/extend head should be allowed to return 0 with change in odp_packet_head(). 

But I guess the odp_packet_data() is the pointer which is allowed to change for trunc/extend head rather than odp_packet_head().

Reply:
> rc should never be zero

This is not actually true. We both misunderstood the API. The should probably say that all pointers to packet data bytes remain valid (meaning, still pointing to the same data bytes), not just that "all pointers remain valid" which can be interpreted, as we did, to mean about all pointers (segment pointer, head pointer, tail pointer etc) that appear in the API.

Linux-gen implementation actually never returns a positive return code from the extend/trunc head functions, which is why the latest version of this patch is now failing in CI tests.

Reply:
@JannePeltonen  Sorry for the delay in updating this PR. I missed your latest comment on this PR.

> (meaning, still pointing to the same data bytes)

In that case , irrespective of the return value rc, pointers like segment pointer, head pointer, tail pointer etc. are to be considered invalid and must not be compared. Right ?

But internal data pointers like L2 ptr, L3 ptr etc. can be compared if rc == 0.

Is my understanding correct ? 


> Linux-gen implementation actually never returns a positive return code from the extend/trunc head functions, which is why the latest version of this patch is now failing in CI tests.

As an experiment, I tried to fix linux-gen to return rc > 0 in case of change in segmentation, but ran into more failures in concat/split test cases. So didn't go down that path further.

Reply:
> In that case , irrespective of the return value rc, pointers like segment pointer, head pointer, tail pointer etc. are to be considered invalid and must not be compared. Right ?

I think so, although the API is perhaps not crystal clear and easy to read.